### PR TITLE
Add functionality for VCF export

### PR DESCRIPTION
**Pull request by christopher-mohr**
With this it is possible to request variants in VCF format by specifying `format=VCF` as query parameter. Additionally it is still possible to get the variants with or without consequences by specifying `withConsequences={true/false}`. The information from the `INFO` column from the VCF the variants were imported from is fetched and exported by default. This could be tricky if you fetch variants from different sources with different `INFO` information and could lead to invalid VCF files (still have to check that).

One feature that is still missing is the export of corresponding genotypes (`withGenotypes=true`) that are available in the database. I guess this should only be possible for variants that share one or multiple genotypes.

### CODE REVIEW

**Comment by christopher-mohr** [comment]
Any feedback is welcome! :)

File: src/main/groovy/life/qbic/oncostore/parser/MetadataContext.groovy [hunk 1]
```diff
@@ -8,13 +8,15 @@ import life.qbic.oncostore.model.Sample
...
-    final Case patient
-    final Sample sample
-    final List<String> vcfFiles
+    boolean isSomatic
+    VariantCaller variantCalling
+    Annotation variantAnnotation
+    ReferenceGenome referenceGenome
+    Case patient
+    Sample sample
+    List<String> vcfFiles
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/oncostore/parser/MetadataContext.groovy:17]: Why did you change the properties to be not final anymore do you expect a change?

File: src/main/groovy/life/qbic/oncostore/model/VcfInfo.groovy [hunk 2]
```diff
@@ -119,4 +122,16 @@ class VcfInfo {
...
     }
+
+    String toVcfFormat() {
+        def vcfInfoString = new StringJoiner(VcfConstants.PROPERTY_DELIMITER)
+        this.properties.each { it ->
+            if (it.key != "class" & it.value != null) {
+                def name = it.key.toString().toUpperCase() as VcfConstants.VcfInfoAbbreviations
+                vcfInfoString.add("${name.getTag()}${VcfConstants.PROPERTY_DEFINITION_STRING}${it.value.toString()}")
+            }
+        }
```
Comment (Steffengreiner) [inline src/main/groovy/life/qbic/oncostore/model/VcfInfo.groovy:133]: I would suggest some form of Exception Handling here concerning vcfInfoString to inform the user if a property is missing.

File: src/main/groovy/life/qbic/oncostore/util/VariantExporter.groovy [hunk 3]
```diff
@@ -10,24 +11,37 @@ class VariantExporter {
...
         variants.each { var ->
             vcfContent.append(var.toVcfFormat())
-            vcfContent.append(var.consequences.collect{AnnotationHandler.toSnpEff(it)}.join(","))
+            if (withConsequences) {
+                vcfContent.append(";")
+                if (metadata.variantAnnotation.name == "snpeff") {
+                    vcfContent.append("${AnnotationHandler.AnnotationTools.SNPEFF.tag}=")
+                    vcfContent.append(var.consequences.collect { AnnotationHandler.toSnpEff(it) }.join(","))
+                }
+                else {
```
Comment (Steffengreiner) [inline src/main/groovy/life/qbic/oncostore/util/VariantExporter.groovy:40]: Is there the possibilty of errors in the database? Right now this code would also add an VEP.tag to an entry with a typo such as "NPEFF". If that is a possibility, I'd suggest explicit checks for both variants and an exception else case to handle possible errors.

**Review by Steffengreiner — COMMENTED** [review]
Looks good to me, I have only 2 suggestions which could also be irrelevant depending on the context of the code execution :)

File: src/main/groovy/life/qbic/oncostore/util/VariantExporter.groovy (hunk 3 above)
Comment (christopher-mohr) [reply src/main/groovy/life/qbic/oncostore/util/VariantExporter.groovy:40]: @Steffengreiner thanks for the feedback! good point! I would say that this should/can be check during import while parsing the metadata. Whats your opinion on that?

Comment (Steffengreiner) [reply src/main/groovy/life/qbic/oncostore/util/VariantExporter.groovy:40]: @christopher-mohr Yes i think that would be the best solution! Good Idea 👍

File: src/main/groovy/life/qbic/oncostore/model/VcfInfo.groovy (hunk 2 above)
Comment (christopher-mohr) [reply src/main/groovy/life/qbic/oncostore/model/VcfInfo.groovy:133]: Do you mean if the value is `null` for any of the defined properties?

Comment (Steffengreiner) [reply src/main/groovy/life/qbic/oncostore/model/VcfInfo.groovy:133]: Yes, although i don't know if that case could occur. 
Right now as a User i wouldn't know the reason why the else case occurs(was it.key == "class" or was it's value == null?). 
I'd suggest a notification that states the property values explicitly in the else case.

File: src/main/groovy/life/qbic/oncostore/parser/MetadataContext.groovy (hunk 1 above)
Comment (christopher-mohr) [reply src/main/groovy/life/qbic/oncostore/parser/MetadataContext.groovy:17]: Thanks for the feedback @KochTobi ! As far as I remember, the reason was that I'm creating a new metadata context in `VariantstoreInformationCenter` during variant export in order to store/provide the reference genome and annotation software information. Would you prefer a constructor with these two properties?

File: src/main/groovy/life/qbic/oncostore/model/VcfInfo.groovy (hunk 2 above)
Comment (christopher-mohr) [reply src/main/groovy/life/qbic/oncostore/model/VcfInfo.groovy:133]: The reason for that is that most of these properties are optional in `VCF` files. Most of them will not be there in most cases. If they were there in the imported `VCF` file, they will be there in the exported one as well...

Comment (Steffengreiner) [reply src/main/groovy/life/qbic/oncostore/model/VcfInfo.groovy:133]: Ah, Thanks for clarifying! Then I don't think it matters since the user can check the exported VCF file

File: src/main/groovy/life/qbic/oncostore/parser/MetadataContext.groovy (hunk 1 above)
Comment (KochTobi) [reply src/main/groovy/life/qbic/oncostore/parser/MetadataContext.groovy:17]: That's a question on what the metadata context is supposed to provide. mandatory metadata should be final but optional could also be non-final. Since in the case of the VariantStoreInformationCenter you only use it to store `referenceGenome` and `annotationSoftware` which you always expect to be there maybe consider changing the method return from `[variants, metadata]` to `[variants, referenceGenomes, annotationSoftware]`. I also noticed that you do only retrieve information from the first variant. Can you be sure that this information is the same across all detected variants found by your query?

Comment (christopher-mohr) [reply src/main/groovy/life/qbic/oncostore/parser/MetadataContext.groovy:17]: I have to think about the first part a bit ;)
>  I also noticed that you do only retrieve information from the first variant. Can you be sure that this information is the same across all detected variants found by your query?

Good point. I did it like that for now but in the end you cannot be sure. It's a bit tricky because theoretically you could have variants in the store that have been annotated with different annotation software and based on different reference genomes. In the end it wouldn't make sense to have these variants in the same exported VCF file. 

Now I'm think if it might be better to make `referenceGenomes` and `annotationSoftware` mandatory query parameters (with default values) for the VCF export.

Comment (KochTobi) [reply src/main/groovy/life/qbic/oncostore/parser/MetadataContext.groovy:17]: That might be a solution for making sure you export the correct variants/information.

Comment (christopher-mohr) [reply src/main/groovy/life/qbic/oncostore/parser/MetadataContext.groovy]: I adapted it accordingly. Would be cool if you can have a look and let me know what you think 👍

File: src/main/groovy/life/qbic/oncostore/controller/VariantController.groovy [hunk 4]
```diff
@@ -90,27 +91,28 @@ class VariantController {
             schema = @Schema(implementation = Variant.class, type = "object")))
     @ApiResponse(responseCode = "400", description = "Invalid variant identifier supplied")
     @ApiResponse(responseCode = "404", description = "Variant not found")
-    HttpResponse<List<Variant>> getVariants(@Nullable ListingArguments args, @Nullable String format, @QueryValue(defaultValue = "false") @Nullable
-            Boolean withConsequences) {
+    HttpResponse<List<Variant>> getVariants(@Nullable ListingArguments args, @Nullable String format, @QueryValue
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/oncostore/controller/VariantController.groovy:86]: The method looks good. But it is not used?

**Review by KochTobi — APPROVED** [review]
Looks good to me

File: src/main/groovy/life/qbic/oncostore/controller/VariantController.groovy (hunk 4 above)
Comment (christopher-mohr) [reply src/main/groovy/life/qbic/oncostore/controller/VariantController.groovy:86]: This is triggered automatically when the corresponding endpoint is called

File: src/main/groovy/life/qbic/oncostore/util/VariantExporter.groovy (hunk 3 above)
Comment (christopher-mohr) [reply src/main/groovy/life/qbic/oncostore/util/VariantExporter.groovy]: @Steffengreiner I added a check for the variant annotation software name (and removed the metadata property VCF files since this is now provided via `files`).

File: src/main/groovy/life/qbic/oncostore/controller/VariantController.groovy [hunk 5]
```diff
@@ -141,12 +141,12 @@ class VariantController {
                             uuid = statusId
                             fileName = file.filename
                             fileSize = file.size
-                            status = Status.processing
+                            status = life.qbic.oncostore.model.Status.processing
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/oncostore/controller/VariantController.groovy:144]: why the prefix? Is there another dependency that introduces a `Status`?

**Review by KochTobi — APPROVED** [review]

File: src/main/groovy/life/qbic/oncostore/controller/VariantController.groovy (hunk 5 above)
Comment (christopher-mohr) [reply src/main/groovy/life/qbic/oncostore/controller/VariantController.groovy:144]: There is also a class `Status` in here: `import io.micronaut.http.annotation.*`

**Comment by christopher-mohr** [comment]
Thanks for your feedback @KochTobi and @Steffengreiner !
